### PR TITLE
3.3

### PR DESCRIPTION
**Pull request by Juniverse**
Main Features:
1. Optimized Auto Reconnection, reducecd connection time.
2. Use tarball to replace github ref.

### CODE REVIEW

File: Runtime/Scripts/UI/BubblePanel.cs [hunk 1]
```diff
@@ -54,7 +54,7 @@ protected virtual void RemoveBubble(string key)
             InworldUIElement elementToDestroy = m_Bubbles[key];
             float height = elementToDestroy.Height;
             m_Bubbles.Remove(key);
-            Destroy(elementToDestroy);
+            DestroyImmediate(elementToDestroy);
```
Comment (mattkwilson) [inline Runtime/Scripts/UI/BubblePanel.cs:57]: Is DestroyImmediate necessary here? If we need to wait for the object to be destroyed before UpdateContent we should use an enumerator.

**Review by mattkwilson — CHANGES_REQUESTED** [review]
Good work!

Some notable points:

- Creating lots of new lists (e.g. `new List<string>{brainName}`) in the new API is going to create lots of garbage
- Push-to-talk needs a revamp, ManualAudioHandling in CharacterHandler is no longer utilized, I noticed when testing audio is still recorded when the PTT button is not held down

File: Runtime/Scripts/Data/Entities/InworldCharacterData.cs [hunk 2]
```diff
@@ -34,6 +34,18 @@ public string CharacterFileName
                 return data.Length < 4 ? brainName : $"{data[3]}_{data[1]}";
             }
         }
+        /// <summary>
+        /// This function is used to rename to fullName. Somehow the current result from new method is naming like aaa__bbb
```
[the remark this answers is not in the document]
Comment (Juniverse) [reply Runtime/Scripts/Data/Entities/InworldCharacterData.cs:38]: Mikhail Fixed this bug, no need to call this func now.

File: Runtime/Scripts/Data/Packets/RelationPacket.cs [hunk 3]
```diff
@@ -76,6 +85,7 @@ public RelationPacket(InworldPacket rhs, RelationEvent evt) : base(rhs)
             type = "RELATION";
             debugInfo = evt;
         }
+        public override string ToJson => JsonUtility.ToJson(this); 
         public string Relation => debugInfo.relation.relationState.GetUpdate(debugInfo.relation.relationUpdate);
```
[the remark this answers is not in the document]
Comment (Juniverse) [reply Runtime/Scripts/Data/Packets/RelationPacket.cs:89]: That's the format of the data returned, the field has to be named as `debugInfo`

File: Runtime/Scripts/Interaction/InworldInteraction.cs [hunk 4]
```diff
@@ -154,10 +151,11 @@ protected IEnumerator HandleNextUtterance()
                 }
                 if (m_CurrentInteraction != null && m_CurrentInteraction.CurrentUtterance != null)
                 {
-                    yield return PlayNextUtterance();
+                    if (InworldController.Audio.SampleMode != MicSampleMode.TURN_BASED || !InworldController.Audio.CurrentPlayingPlayer || !InworldController.Audio.CurrentPlayingPlayer.isPlaying)
```
[the remark this answers is not in the document]
Comment (Juniverse) [reply Runtime/Scripts/Interaction/InworldInteraction.cs:154]: Use `CurrentPlayingAudioSource`

**Review by mattkwilson — APPROVED** [review]

File: Runtime/Scripts/UI/BubblePanel.cs (hunk 1 above)
Comment (Juniverse) [reply Runtime/Scripts/UI/BubblePanel.cs]: change back to Destroy()

**Pull request merged**